src/ros_filter.cpp

```diff
@@ -2894,32 +2964,56 @@ namespace RobotLocalization
-    state->latestControl_ = Eigen::VectorXd(filter.getControl());
-    state->latestControlTime_ = filter.getControlTime();
-    filterStateHistory_.push_back(state);
-    RF_DEBUG("Saved state with timestamp " << std::setprecision(20) << state->lastMeasurementTime_ <<
-             " to history. " << filterStateHistory_.size() << " measurements are in the queue.\n");
-  }
-
-  template<typename T>
   bool RosFilter<T>::revertTo(const double time)
   {
     RF_DEBUG("\n----- RosFilter::revertTo -----\n");
-    RF_DEBUG("\nRequested time was " << std::setprecision(20) << time << "\n")
+    RF_DEBUG("\nRequested time was " << std::setprecision(20) << time << "\n");
 
-    // Walk back through the queue until we reach a filter state whose time stamp is less than or equal to the requested time.
-    // Since every saved state after that time will be overwritten/corrected, we can pop from the queue.
+    // Walk back through the queue until we reach a filter state whose time stamp is less than or equal to the
+    // requested time. Since every saved state after that time will be overwritten/corrected, we can pop from
+    // the queue.
     while (!filterStateHistory_.empty() && filterStateHistory_.back()->lastMeasurementTime_ > time)
     {
       filterStateHistory_.pop_back();
     }
 
+    // If we jump back in the state history, then we unfortunately need, for each differential input source, to find
+    // the *previous* filter state for that measurement (i.e., the filter state at the *previous* measurement for that
+    // sensor).
+
+    // Iterate over all the previousMeasurementStates_, since we will only have entries for the differential sensors
+    std::map<std::string, FilterStatePtr>::iterator pmsIt;
+    std::map<std::string, FilterStatePtr>::iterator pmsErased;
+    for (pmsIt = previousMeasurementStates_.begin(); pmsIt != previousMeasurementStates_.end();)
+    {
+      // Walk backwards through the history until we find a state that resulted from a message with this topic name
+      FilterStateHistoryDeque::reverse_iterator fshIt = filterStateHistory_.rbegin();
+      while (fshIt != filterStateHistory_.rend() && (*fshIt)->associatedTopicName_ != pmsIt->first)
+      {
+        fshIt++;
+      }
+
+      // If we found it, revert the previous state for that topic name
+      if (fshIt != filterStateHistory_.rend())
+      {
+        pmsIt->second = (*fshIt);
+
+        RF_DEBUG("Reverting previous state for " << pmsIt->first << " to:\n" << pmsIt->second->state_ <<
+                 "Reverting previous covariance for " << pmsIt->first << " to:\n" << pmsIt->second->estimateErrorCovariance_);
+
+        ++pmsIt;
+      }
+      else
+      {
+        RF_DEBUG("Erasing previous measurement for " << pmsIt->first);
+
+        // If we failed to find it, erase this sensor from the previous states
+        pmsErased = pmsIt;
+        ++pmsIt;
+        previousMeasurementStates_.erase(pmsErased);
```

Comment:
If the purpose of `pmsErased` is solely to avoid iterator invalidation, you can normally do this:
```pmsIt = previousMeasurementStates_.erase(pmsIt);```
The `erase()` method returns an iterator to the item after the just-erased item.

Reply:
That doesn't work with maps in C++98:

http://www.cplusplus.com/reference/map/map/erase/

(Click on the C++98 side tab). It's an old `std::map` issue that was fixed in C++11.

Reply:
The other trick is using the post-increment operator: `previousMeasurementStates_.erase(pmsIt++);`

Reply:
Will the order play ball there? As in, will the compiler increment the iterator before executing the function?

Reply:
The X++ operator makes a copy of the original value, increments the value, then returns the copy to the `erase()` function. The internal copy is why you shouldn't use it most of the time.

Reply:
Sample implementation for some custom iterator on SO:
```
Iterator<T> operator++(int) 
{ 
    node<T>* previous = pointer;
    pointer = pointer->next; 
    return Iterator<T>(previous);
}
```

Reply:
Yep, updated it to the postfix operator.